### DISCUSSION TITLE
Prefer Func<Task> over Func<ValueTask> overload when using async lambdas

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/csharplang#4360] [Q&A] [answered]
I have one method with two overloads:
```cs
void Foo(Func<Task> task)
void Foo(Func<ValueTask> task)
```

All works well until async lambda is passed as a first argument: `Foo(async () => {})`. Since lambda itself doesn't have a type, compiler doesn't know what overload I'm trying to invoke and it gives [the ambiguity error](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0121). I can use explicit cast to resolve it (`Foo((Func<Task>)(async () => {}))`), but I don't want to force users of `Foo` to write it every time in this situation.

Can I somehow set `void Foo(Func<Task> task)` as a tiebreaker? So that there is no compile errors when user writes this `Foo(async () => {})`

**Comment by YairHalberstadt** [comment, marked as answer]
I did it by using a dummy optional parameter:

https://github.com/YairHalberstadt/stronginject/blob/86f0e766bb2a2028d302f9402db166ca42cb7b84/StrongInject/ContainerExtensions.cs#L44-L56

So in your case that would be:

```csharp

/// <summary>
/// A class with no possible value other than null. Used to mark an optional parameter which should never be set.
/// </summary>
[EditorBrowsable(EditorBrowsableState.Never)]
public sealed class DummyParameter
{
    private DummyParameter() { }
}
    
void Foo(Func<Task> task)
void Foo(Func<ValueTask> task, DummyParameter? _ = null)
```

**Reply by maxkoshevoi** [reply]
Thank you so much! That helped a lot. 
I used your advice in reverse. My methods have lots of optional parameters, so I created overloads for `void Foo(Func<Task> task)` without them.

https://github.com/xamarin/XamarinCommunityToolkit/pull/797/commits/ca6b60690155af2b4b569356eb9b93f56e15411b

**Comment by huoyaoyuan** [comment]
Another idea: use different names and named argument?

```csharp
void Foo(Func<Task> taskFunc);
void Foo(Func<ValueTask> valueTaskFunc);

Foo(taskFunc: () => ...)
```

(haven't tried it, but I think it could work)

**Reply by maxkoshevoi** [reply]
Thank you! This also will work, but I think it's a little unobvious to the user (and would require some work from them). Implemented both solutions in my code with this as a backup one.
